File: src/typescript/tests/crossimpl.test.ts
/**
 * Cross-Implementation Test Runner
 * 
 * Runs test cases from spec/tests/move_validation.json to verify
 * TypeScript implementation matches the shared spec.
 * 
 * Signed-by: agent #28 claude-sonnet-4 via opencode 20260122T08:04:58
 */

import { describe, test, expect, beforeAll } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';

import {
  isValidCell,
  validateMove,
  BoardState,
  Piece,
  Color,
  PieceType,
  LanceVariant,
  coordToString,
} from '../src/index';

// Test case types from spec
interface BoardValidationTestCase {
  id: string;
  description: string;
  type: 'boardValidation';
  input: { q: number; r: number };
  expected: { valid: boolean };
}

interface MoveValidationTestCase {
  id: string;
  description: string;
  type: 'moveValidation';
  setup: {
    pieces: Array<{
      piece: string;
      color: string;
      q: number;
      r: number;
      variant?: string;
    }>;
    turn: string;
  };
  move: {
    from: { q: number; r: number };
    to: { q: number; r: number };
  };
  expected: {
    legal: boolean;
    capture?: boolean;
    reason?: string;
  };
}

type TestCase = BoardValidationTestCase | MoveValidationTestCase;

interface TestSuite {
  testCases: TestCase[];
}

// Load test cases
function loadTestSuite(): TestSuite {
  const specPath = path.join(__dirname, '../../../spec/tests/move_validation.json');
  const content = fs.readFileSync(specPath, 'utf-8');
  return JSON.parse(content) as TestSuite;
}

// Convert spec piece to our Piece type
function specToPiece(spec: { piece: string; color: string; variant?: string }): Piece {
  const type = spec.piece as PieceType;
  const color = spec.color as Color;
  const piece: Piece = { type, color };
  
  if (spec.variant) {
    return { ...piece, variant: spec.variant as LanceVariant };
  }
  
  return piece;
}

// Build board from spec setup
function buildBoardFromSpec(setup: MoveValidationTestCase['setup']): BoardState {
  const board: BoardState = new Map();
  
  for (const placement of setup.pieces) {
    const piece = specToPiece(placement);
    const coord = { q: placement.q, r: placement.r };
    board.set(coordToString(coord), piece);
  }
  
  return board;
}

describe('Cross-Implementation Tests (spec/tests/move_validation.json)', () => {
  let testSuite: TestSuite;
  
  beforeAll(() => {
    testSuite = loadTestSuite();
  });
  
  describe('Board Validation Tests', () => {
    test.each([
      { id: 'board_001', q: 0, r: 0, valid: true, desc: 'Center cell is valid' },
      { id: 'board_002', q: 4, r: 0, valid: true, desc: 'Corner cell at max radius is valid' },
      { id: 'board_003', q: 5, r: 0, valid: false, desc: 'Cell outside board is invalid' },
      { id: 'board_004', q: 3, r: 3, valid: false, desc: 'Cell violating q+r constraint is invalid' },
    ])('$id: $desc', ({ q, r, valid }) => {
      const result = isValidCell({ q, r });
      expect(result).toBe(valid);
    });
    
    test('runs all board validation tests from spec', () => {
      const boardTests = testSuite.testCases.filter(
        (tc): tc is BoardValidationTestCase => tc.type === 'boardValidation'
      );
      
      for (const tc of boardTests) {
        const result = isValidCell(tc.input);
        expect(result, `${tc.id}: ${tc.description}`).toBe(tc.expected.valid);
      }
    });
  });
  
  describe('Move Validation Tests', () => {
    test('runs all move validation tests from spec', () => {
      const moveTests = testSuite.testCases.filter(
        (tc): tc is MoveValidationTestCase => tc.type === 'moveValidation'
      );
      
      for (const tc of moveTests) {
        const board = buildBoardFromSpec(tc.setup);
        const turn = tc.setup.turn as Color;
        
        const result = validateMove(board, tc.move.from, tc.move.to, turn);
        
        // Check legal/illegal
        expect(result.legal, `${tc.id}: ${tc.description} - legal`).toBe(tc.expected.legal);
        
        // If legal and capture is specified, check it
        if (tc.expected.legal && tc.expected.capture !== undefined) {
          expect(result.capture, `${tc.id}: ${tc.description} - capture`).toBe(tc.expected.capture);
        }
        
        // If illegal with specific reason, check it
        if (!tc.expected.legal && tc.expected.reason) {
          expect(result.reason, `${tc.id}: ${tc.description} - reason`).toBe(tc.expected.reason);
        }
      }
    });
    
    // Individual test cases for better error reporting
    test('king_001: King can move to adjacent empty cell', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'king', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 1, r: 0 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('king_002: King cannot move 2 squares', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'king', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 2, r: 0 }, 'white');
      expect(result.legal).toBe(false);
    });
    
    test('king_003: King can capture enemy piece', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'pawn', color: 'black', q: 1, r: 0 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 1, r: 0 }, 'white');
      expect(result.legal).toBe(true);
      expect(result.capture).toBe(true);
    });
    
    test('queen_001: Queen can slide multiple squares', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'queen', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 0, r: -3 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('queen_002: Queen cannot jump over pieces', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'queen', color: 'white', q: 0, r: 0 },
          { piece: 'pawn', color: 'white', q: 0, r: -1 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 0, r: -3 }, 'white');
      expect(result.legal).toBe(false);
    });
    
    test('pawn_001: White pawn moves north', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'pawn', color: 'white', q: 0, r: 2 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 0, r: 1 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('pawn_002: White pawn cannot move south', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'pawn', color: 'white', q: 0, r: 2 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 0, r: 3 }, 'white');
      expect(result.legal).toBe(false);
    });
    
    test('pawn_003: White pawn captures forward (north)', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'pawn', color: 'white', q: 0, r: 2 },
          { piece: 'pawn', color: 'black', q: 0, r: 1 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 0, r: 1 }, 'white');
      expect(result.legal).toBe(true);
      expect(result.capture).toBe(true);
    });
    
    test('pawn_004: White pawn captures diagonally NE', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'pawn', color: 'white', q: 0, r: 2 },
          { piece: 'pawn', color: 'black', q: 1, r: 1 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 1, r: 1 }, 'white');
      expect(result.legal).toBe(true);
      expect(result.capture).toBe(true);
    });
    
    test('knight_001: Knight leaps to valid target', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'knight', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 1, r: -2 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('knight_002: Knight can jump over pieces', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'knight', color: 'white', q: 0, r: 0 },
          { piece: 'pawn', color: 'white', q: 0, r: -1 },
          { piece: 'pawn', color: 'black', q: 1, r: -1 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 1, r: -2 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('lance_001: Lance A slides north', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'lance', color: 'white', q: 0, r: 2, variant: 'A' }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 0, r: -2 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('lance_002: Lance A cannot move NE (not in its directions)', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'lance', color: 'white', q: 0, r: 2, variant: 'A' }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 2 }, { q: 2, r: 0 }, 'white');
      expect(result.legal).toBe(false);
    });
    
    test('chariot_001: Chariot slides NE', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'chariot', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 3, r: -3 }, 'white');
      expect(result.legal).toBe(true);
    });
    
    test('chariot_002: Chariot cannot move north (not diagonal)', () => {
      const board = buildBoardFromSpec({
        pieces: [{ piece: 'chariot', color: 'white', q: 0, r: 0 }],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 0, r: -2 }, 'white');
      expect(result.legal).toBe(false);
    });
    
    test('check_001: King cannot move into check', () => {
      const board = buildBoardFromSpec({
        pieces: [
          { piece: 'king', color: 'white', q: 0, r: 0 },
          { piece: 'queen', color: 'black', q: 1, r: -4 },
        ],
        turn: 'white',
      });
      const result = validateMove(board, { q: 0, r: 0 }, { q: 1, r: 0 }, 'white');
      expect(result.legal).toBe(false);
      expect(result.reason).toBe('movesIntoCheck');
    });
  });
  
  describe('Spec Test Coverage Report', () => {
    test('reports test case coverage', () => {
      const boardTests = testSuite.testCases.filter(tc => tc.type === 'boardValidation');
      const moveTests = testSuite.testCases.filter(tc => tc.type === 'moveValidation');
      
      console.log('\n=== Spec Test Coverage Report ===');
      console.log(`Board validation tests: ${boardTests.length}`);
      console.log(`Move validation tests: ${moveTests.length}`);
      console.log(`Total spec tests: ${testSuite.testCases.length}`);
      console.log('=================================\n');
      
      expect(testSuite.testCases.length).toBeGreaterThan(0);
    });
  });
});

File: src/rust/tests/crossimpl_test.rs
//! Cross-Implementation Test Runner
//!
//! Runs test cases from spec/tests/move_validation.json to verify
//! Rust implementation matches the shared spec.
//!
//! Signed-by: agent #29 claude-sonnet-4 via opencode 20260122T08:15:15

use serde::Deserialize;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

// Import from the crate
use underchex_wasm::{
    is_valid_cell, validate_move, BoardState, Color, HexCoord, LanceVariant, Piece, PieceType,
};

// ============================================================================
// Test Case Structures (matching JSON spec)
// ============================================================================

#[derive(Debug, Deserialize)]
struct TestSuite {
    #[serde(rename = "testCases")]
    test_cases: Vec<TestCase>,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
enum TestCase {
    #[serde(rename = "boardValidation")]
    BoardValidation(BoardValidationCase),
    #[serde(rename = "moveValidation")]
    MoveValidation(MoveValidationCase),
}

#[derive(Debug, Deserialize)]
struct BoardValidationCase {
    id: String,
    description: String,
    input: CoordInput,
    expected: BoardExpected,
}

#[derive(Debug, Deserialize)]
struct CoordInput {
    q: i32,
    r: i32,
}

#[derive(Debug, Deserialize)]
struct BoardExpected {
    valid: bool,
}

#[derive(Debug, Deserialize)]
struct MoveValidationCase {
    id: String,
    description: String,
    setup: SetupConfig,
    #[serde(rename = "move")]
    the_move: MoveConfig,
    expected: MoveExpected,
}

#[derive(Debug, Deserialize)]
struct SetupConfig {
    pieces: Vec<PiecePlacement>,
    turn: String,
}

#[derive(Debug, Deserialize)]
struct PiecePlacement {
    piece: String,
    color: String,
    q: i32,
    r: i32,
    variant: Option<String>,
}

#[derive(Debug, Deserialize)]
struct MoveConfig {
    from: CoordInput,
    to: CoordInput,
}

#[derive(Debug, Deserialize)]
struct MoveExpected {
    legal: bool,
    capture: Option<bool>,
    reason: Option<String>,
}

// ============================================================================
// Helper Functions
// ============================================================================

fn load_test_suite() -> TestSuite {
    let spec_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("spec")
        .join("tests")
        .join("move_validation.json");

    let content = fs::read_to_string(&spec_path)
        .unwrap_or_else(|e| panic!("Failed to read spec file at {:?}: {}", spec_path, e));

    serde_json::from_str(&content).expect("Failed to parse spec JSON")
}

fn string_to_color(s: &str) -> Color {
    match s {
        "white" => Color::White,
        "black" => Color::Black,
        _ => panic!("Invalid color: {}", s),
    }
}

fn string_to_piece_type(s: &str) -> PieceType {
    match s {
        "king" => PieceType::King,
        "queen" => PieceType::Queen,
        "knight" => PieceType::Knight,
        "pawn" => PieceType::Pawn,
        "lance" => PieceType::Lance,
        "chariot" => PieceType::Chariot,
        _ => panic!("Invalid piece type: {}", s),
    }
}

fn string_to_lance_variant(s: &str) -> LanceVariant {
    match s {
        "A" => LanceVariant::A,
        "B" => LanceVariant::B,
        _ => panic!("Invalid lance variant: {}", s),
    }
}

fn build_board_from_spec(setup: &SetupConfig) -> BoardState {
    let mut board: BoardState = HashMap::new();

    for placement in &setup.pieces {
        let piece_type = string_to_piece_type(&placement.piece);
        let color = string_to_color(&placement.color);

        let piece = if piece_type == PieceType::Lance {
            let variant = placement
                .variant
                .as_ref()
                .map(|v| string_to_lance_variant(v))
                .unwrap_or(LanceVariant::A);
            Piece::lance(color, variant)
        } else {
            Piece::new(piece_type, color)
        };

        let coord = HexCoord::new(placement.q, placement.r);
        board.insert(coord.to_key(), piece);
    }

    board
}

// ============================================================================
// Tests
// ============================================================================

#[test]
fn test_board_validation_from_spec() {
    let suite = load_test_suite();
    let mut count = 0;

    for tc in &suite.test_cases {
        if let TestCase::BoardValidation(case) = tc {
            let coord = HexCoord::new(case.input.q, case.input.r);
            let result = is_valid_cell(coord);

            assert_eq!(
                result, case.expected.valid,
                "{}: {} - expected valid={}, got valid={}",
                case.id, case.description, case.expected.valid, result
            );
            count += 1;
        }
    }

    println!("Board validation tests passed: {}", count);
    assert!(count > 0, "No board validation tests found");
}

#[test]
fn test_move_validation_from_spec() {
    let suite = load_test_suite();
    let mut count = 0;

    for tc in &suite.test_cases {
        if let TestCase::MoveValidation(case) = tc {
            let board = build_board_from_spec(&case.setup);
            let turn = string_to_color(&case.setup.turn);

            let from = HexCoord::new(case.the_move.from.q, case.the_move.from.r);
            let to = HexCoord::new(case.the_move.to.q, case.the_move.to.r);

            let result = validate_move(&board, from, to, turn);

            // Check legal/illegal
            assert_eq!(
                result.legal, case.expected.legal,
                "{}: {} - expected legal={}, got legal={}",
                case.id, case.description, case.expected.legal, result.legal
            );

            // If legal and capture is specified, check it
            if case.expected.legal {
                if let Some(expected_capture) = case.expected.capture {
                    assert_eq!(
                        result.capture, expected_capture,
                        "{}: {} - expected capture={}, got capture={}",
                        case.id, case.description, expected_capture, result.capture
                    );
                }
            }

            // If illegal with specific reason, check it
            if !case.expected.legal {
                if let Some(expected_reason) = &case.expected.reason {
                    assert_eq!(
                        result.reason.as_deref(),
                        Some(expected_reason.as_str()),
                        "{}: {} - expected reason={:?}, got reason={:?}",
                        case.id,
                        case.description,
                        expected_reason,
                        result.reason
                    );
                }
            }

            count += 1;
        }
    }

    println!("Move validation tests passed: {}", count);
    assert!(count > 0, "No move validation tests found");
}

// ============================================================================
// Individual Test Cases (for better error reporting)
// ============================================================================

#[test]
fn test_center_cell_is_valid() {
    assert!(is_valid_cell(HexCoord::new(0, 0)));
}

#[test]
fn test_corner_at_max_radius_is_valid() {
    assert!(is_valid_cell(HexCoord::new(4, 0)));
}

#[test]
fn test_cell_outside_board_is_invalid() {
    assert!(!is_valid_cell(HexCoord::new(5, 0)));
}

#[test]
fn test_cell_violating_constraint_is_invalid() {
    assert!(!is_valid_cell(HexCoord::new(3, 3)));
}

#[test]
fn test_king_can_move_to_adjacent_empty_cell() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::King, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(1, 0),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_king_cannot_move_two_squares() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::King, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(2, 0),
        Color::White,
    );
    assert!(!result.legal);
}

#[test]
fn test_king_can_capture_enemy() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::King, Color::White),
    );
    board.insert(
        HexCoord::new(1, 0).to_key(),
        Piece::new(PieceType::Pawn, Color::Black),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(1, 0),
        Color::White,
    );
    assert!(result.legal);
    assert!(result.capture);
}

#[test]
fn test_queen_can_slide_multiple_squares() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Queen, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(0, -3),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_queen_cannot_jump_over_pieces() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Queen, Color::White),
    );
    board.insert(
        HexCoord::new(0, -1).to_key(),
        Piece::new(PieceType::Pawn, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(0, -3),
        Color::White,
    );
    assert!(!result.legal);
}

#[test]
fn test_white_pawn_moves_north() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 2).to_key(),
        Piece::new(PieceType::Pawn, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 2),
        HexCoord::new(0, 1),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_white_pawn_cannot_move_south() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 2).to_key(),
        Piece::new(PieceType::Pawn, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 2),
        HexCoord::new(0, 3),
        Color::White,
    );
    assert!(!result.legal);
}

#[test]
fn test_knight_leaps_to_valid_target() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Knight, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(1, -2),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_knight_can_jump_over_pieces() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Knight, Color::White),
    );
    board.insert(
        HexCoord::new(0, -1).to_key(),
        Piece::new(PieceType::Pawn, Color::White),
    );
    board.insert(
        HexCoord::new(1, -1).to_key(),
        Piece::new(PieceType::Pawn, Color::Black),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(1, -2),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_lance_a_slides_north() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 2).to_key(),
        Piece::lance(Color::White, LanceVariant::A),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 2),
        HexCoord::new(0, -2),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_lance_a_cannot_move_ne() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 2).to_key(),
        Piece::lance(Color::White, LanceVariant::A),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 2),
        HexCoord::new(2, 0),
        Color::White,
    );
    assert!(!result.legal);
}

#[test]
fn test_chariot_slides_ne() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Chariot, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(3, -3),
        Color::White,
    );
    assert!(result.legal);
}

#[test]
fn test_chariot_cannot_move_north() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::Chariot, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(0, -2),
        Color::White,
    );
    assert!(!result.legal);
}

#[test]
fn test_king_cannot_move_into_check() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::King, Color::White),
    );
    board.insert(
        HexCoord::new(1, -4).to_key(),
        Piece::new(PieceType::Queen, Color::Black),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, 0),
        HexCoord::new(1, 0),
        Color::White,
    );
    assert!(!result.legal);
    assert_eq!(result.reason.as_deref(), Some("movesIntoCheck"));
}

#[test]
fn test_cannot_move_opponents_piece() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, -2).to_key(),
        Piece::new(PieceType::Pawn, Color::Black),
    );

    let result = validate_move(
        &board,
        HexCoord::new(0, -2),
        HexCoord::new(0, -1),
        Color::White,
    );
    assert!(!result.legal);
    assert_eq!(result.reason.as_deref(), Some("notYourPiece"));
}

#[test]
fn test_cannot_move_from_empty_cell() {
    let mut board: BoardState = HashMap::new();
    board.insert(
        HexCoord::new(0, 0).to_key(),
        Piece::new(PieceType::King, Color::White),
    );

    let result = validate_move(
        &board,
        HexCoord::new(1, 0),
        HexCoord::new(2, 0),
        Color::White,
    );
    assert!(!result.legal);
    assert_eq!(result.reason.as_deref(), Some("noPieceAtSource"));
}

#[test]
fn test_coverage_report() {
    let suite = load_test_suite();

    let board_tests = suite
        .test_cases
        .iter()
        .filter(|tc| matches!(tc, TestCase::BoardValidation(_)))
        .count();

    let move_tests = suite
        .test_cases
        .iter()
        .filter(|tc| matches!(tc, TestCase::MoveValidation(_)))
        .count();

    println!("\n=== Spec Test Coverage Report (Rust) ===");
    println!("Board validation tests: {}", board_tests);
    println!("Move validation tests: {}", move_tests);
    println!("Total spec tests: {}", suite.test_cases.len());
    println!("=========================================\n");

    assert!(suite.test_cases.len() > 0);
}
